### ISSUE TITLE
Request: output file flag for `sonobuoy retrieve`

### ISSUE THREAD

**Issue by stone-z** [vmware-tanzu/sonobuoy#1361]
**Describe the solution you'd like**

TL;DR: It would be helpful to have an `--output` flag for `sonobuoy retrieve` specifying a path to save the results.

Background:
With the move to `distroless`, there is no longer a shell in the container, which has made it a little harder to use the official `sonobuoy` Docker images in our CI/CD pipelines.

Up until now, we had [this Tekton Task](https://github.com/giantswarm/test-infra/blob/master/tekton/tasks/manage-sonobuoy-results.yaml) which would run `sonobuoy retrieve`, leave the results in a known path for the next CI step, and clean up the namespace. Now there's not a great way to get the results back from the official image, and we'd prefer to stick to the official images and not maintain our own (not to mention we'd rather use the distroless images anyway).

**Anything else you would like to add:**
Alternative solutions are also welcome if there are better practices for Docker image usage than I was able to find in the docs.

**Environment:**

- Sonobuoy version: v0.52.0
- Kubernetes version: (use `kubectl version`): 1.20.8
- Kubernetes installer & version:
- Cloud provider or hardware configuration: AWS via Giant Swarm
- OS (e.g. from `/etc/os-release`): Ubuntu 20.04 (client), Flatcar 2765.2.6 (server)

**Comment by johnSchnake** [comment]
So the flag you're suggesting is effectively to remove the need for this specific line: https://github.com/giantswarm/test-infra/blob/master/tekton/tasks/manage-sonobuoy-results.yaml#L41 is that correct?

It seems to me that this is already supported but maybe I am not fully understanding the issue. Were you aware that `retrieve` takes an optional argument for the location (and also has a flag for automatically extracting the contents)?

```
$ sonobuoy retrieve /Users/jschnake/abspath-is-ok
/Users/jschnake/abspath-is-ok/202107201455_sonobuoy_e9de5e09-bdcc-44eb-bda7-5ac3507b2c87.tar.gz
```

or 

```
$ sonobuoy retrieve /Users/jschnake/abspath-is-ok --extract
$ ls -lah ~/abspath-is-ok
total 24
drwxr-xr-x   9 jschnake  staff   288B Jul 20 09:59 .
drwxr-xr-x+ 88 jschnake  staff   2.8K Jul 20 09:59 ..
drwxr-xr-x   5 jschnake  staff   160B Jul 20 09:59 hosts
drwxr-xr-x   6 jschnake  staff   192B Jul 20 09:59 meta
drwxr-xr-x   4 jschnake  staff   128B Jul 20 09:59 plugins
drwxr-xr-x   3 jschnake  staff    96B Jul 20 09:59 podlogs
drwxr-xr-x   4 jschnake  staff   128B Jul 20 09:59 resources
-rw-r--r--   1 jschnake  staff   4.6K Jul 20 09:59 servergroups.json
-rw-r--r--   1 jschnake  staff   226B Jul 20 09:59 serverversion.json
```

**Comment by stone-z** [comment]
> So the flag you're suggesting is effectively to remove the need for this specific line: https://github.com/giantswarm/test-infra/blob/master/tekton/tasks/manage-sonobuoy-results.yaml#L41 is that correct?

That is basically the root of it, yeah. I must have missed in the docs that it was possible to specify the output folder path, but after trying that route it seems I'm still stuck because the name of the actual archive is still different between runs.

Essentially what we have now combines `status`, `retrieve`, and `results` in one CI step, so that it is human readable, we can detect when the run fails, and then push the detailed results to an S3 bucket in a later task. Using individual sonobuoy commands in separate tasks using the distroless image we can't really reproduce this directly because there is no way to manipulate the sonobuoy output. We can run `status` and `retrieve`, but would have to pass the archive name out of the `retrieve` step (without bash, so parsing the container output from that step) and pass it into any future steps if we want to operate on the results later (e.g. to use `sonobuoy results`) and still package the "did this run fail or not?" logic separately.

So after playing with this for a bit, I think we're just going to have to maintain an alpine image. It seems more straightforward than having separate CI steps for `run`, `status`, `retrieve`, `results`, and then still needing some bash to handle the business logic.

Specifying the exact output name would still be helpful so that CI/CD steps don't have to pass the filename forward, but we'll go a different route so this isn't a blocker.

**Comment by johnSchnake** [comment]
OK thanks for expanding on your issue. I'm going to have a think about it; I'm out of the office until Thursday so I'll figure something out and comment again soon. I definitely don't want you to have to maintain a separate image just because it is hard to script around without relying on bash.

**Comment by stone-z** [comment]
Sure thing, thanks for taking a look. We aren't blocked here so no rush on our side. I also appreciate the `distroless` move generally, we are just limited by Tekton in this case.

For others finding this and wanting an alpine image, we have a public one on [Quay](https://quay.io/repository/giantswarm/sonobuoy) and [Docker Hub](https://hub.docker.com/r/giantswarm/sonobuoy). It is [built and pushed automatically](https://github.com/giantswarm/retagger/blob/9775d28e154bf18b4b6663aa85b712e92488a931/images.yaml#L826) for each new `sonobuoy` tagged image.

**Comment by johnSchnake** [comment]
I'm going to put in this feature today and hopefully merge/release afterward. I think this is a simple problem we can solve for you. Thanks for bringing it to our attention.

**Comment by stone-z** [comment]
Thank you!
